File: grind75/l0383/ransom-notes.py
from collections import defaultdict


def canConstruct(ransomNote: str, magazine: str) -> bool:
    hm = defaultdict(int)
    for ch in magazine:
        hm[ch] += 1
    
    for ch in ransomNote:
        if ch not in hm or hm[ch] == 0:
            return False
        else:
            hm[ch] -= 1
    
    return True

File: grind75/l0383/RansomNotes.java
package grind75.l0383;


import java.util.HashMap;
import java.util.Map;

public class RansomNotes {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> hm = new HashMap<>();
        for (Character ch: magazine.toCharArray()) {
            int cur = hm.getOrDefault(ch, 0);
            hm.put(ch, cur + 1);
        }

        for (Character ch: ransomNote.toCharArray()) {
            if (!hm.containsKey(ch) || hm.get(ch) == 0) {
                return false;
            }
            else {
                hm.put(ch, hm.get(ch)-1);
            }
        }
        return true;
    }
}
    

